src/main/java/com/timgroup/statsd/StatsDSender.java

```diff
@@ -5,61 +5,80 @@
 import java.nio.ByteBuffer;
 import java.nio.channels.DatagramChannel;
 import java.nio.charset.Charset;
-import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.Callable;
-import java.util.concurrent.LinkedBlockingQueue;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.TimeUnit;
 
 public class StatsDSender implements Runnable {
     private static final Charset MESSAGE_CHARSET = Charset.forName("UTF-8");
     private static final String MESSAGE_TOO_LONG = "Message longer than size of sendBuffer";
+    private static final int WAIT_SLEEP_MS = 10;  // 10 ms would be a 100HZ slice
 
     private final ByteBuffer sendBuffer;
     private final Callable<SocketAddress> addressLookup;
-    private final BlockingQueue<String> queue;
+    private final SocketAddress address;
     private final StatsDClientErrorHandler handler;
     private final DatagramChannel clientChannel;
 
+    private final int qCapacity;
+    private final Queue<String> queue;
+    private final AtomicInteger qSize;  // qSize will not reflect actual size, but a close estimate.
+
     private volatile boolean shutdown;
 
 
     StatsDSender(final Callable<SocketAddress> addressLookup, final int queueSize,
-                 final StatsDClientErrorHandler handler, final DatagramChannel clientChannel, final int maxPacketSizeBytes) {
-        this(addressLookup,  new LinkedBlockingQueue<String>(queueSize), handler, clientChannel, maxPacketSizeBytes);
-    }
+                 final StatsDClientErrorHandler handler, final DatagramChannel clientChannel, final int maxPacketSizeBytes)
+            throws Exception {
+
+        this.qSize = new AtomicInteger(0);
+        this.qCapacity = queueSize;
 
-    StatsDSender(final Callable<SocketAddress> addressLookup, final BlockingQueue<String> queue,
-                 final StatsDClientErrorHandler handler, final DatagramChannel clientChannel, final int maxPacketSizeBytes) {
-        sendBuffer = ByteBuffer.allocate(maxPacketSizeBytes);
         this.addressLookup = addressLookup;
-        this.queue = queue;
+        this.address = addressLookup.call();
+        this.queue = new ConcurrentLinkedQueue<String>();
         this.handler = handler;
         this.clientChannel = clientChannel;
-    }
 
+        sendBuffer = ByteBuffer.allocateDirect(maxPacketSizeBytes);
+    }
 
     boolean send(final String message) {
         if (!shutdown) {
-            queue.offer(message);
-            return true;
+            if (qSize.get() < qCapacity) {
+                queue.offer(message);
```

Comment:
The signature for `offer` is `boolean java.util.Queue.offer(String e)`.  I do not know if the function can return false for `ConcurrentLinkedQueue` but if it is possible, `qSize.get()` may not be correct. This could lead to a situation where the queue is empty but `qSize.get() >= qCapacity`.

Reply:
It's true that `qCapacity` is not a fully accurate picture. I resorted to this "soft guide" to set bounds on the queue size, but exceeding the bounds somewhat wasn't a problem. Since for every push there **will** be a pop, it didn't seem like much of a problem. We also cannot pop if there isn't an element in the queue, so the value of `qSize` could never be negative. We could however exceed the value for `qCapacity` in which case we'd drop so the value would be guaranteed to come back down as we `pop()` in the processing routine. I guess what I'm trying to say is that even though the value is not accurate, it still allows us to set a vaguely accurate bound on the queue size. 

Finally, `offer()` cannot return false for a ConcurrentLinkedQueue. From the javadoc:
> As the queue is unbounded, this method will never return false.

Let me know if you see any holes in my reasoning.